src/Compilers/CSharp/Portable/Symbols/Source/SourceTypeParameterSymbol.cs

```diff
@@ -217,18 +228,25 @@ internal override void EnsureAllConstraintsAreResolved(bool canIgnoreNullableCon
             get;
         }
 
-        private TypeParameterBounds GetBounds(ConsList<TypeParameterSymbol> inProgress, bool canIgnoreNullableContext)
+        private TypeParameterBounds GetBounds(ConsList<TypeParameterSymbol> inProgress, bool canUseLightweightTypeConstraintBinding)
         {
             Debug.Assert(!inProgress.ContainsReference(this));
             Debug.Assert(!inProgress.Any() || ReferenceEquals(inProgress.Head.ContainingSymbol, this.ContainingSymbol));
 
-            if (!_lazyBounds.HasValue(canIgnoreNullableContext))
+            if (!_lazyBounds.HasValue(canUseLightweightTypeConstraintBinding))
             {
                 var diagnostics = DiagnosticBag.GetInstance();
-                var bounds = this.ResolveBounds(inProgress, canIgnoreNullableContext, diagnostics);
+                var bounds = this.ResolveBounds(inProgress, canUseLightweightTypeConstraintBinding, diagnostics);
+
+                if (bounds is null && canUseLightweightTypeConstraintBinding)
```

Comment:
>if (bounds is null && canUseLightweightTypeConstraintBinding) [](start = 16, length = 61)

If there were no constraint clauses, can we just store `null`? Can the decision of whether to use `null` or `NullFromLightweightBinding` be handled in `ConstraintsHelper.ResolveBounds()` rather than here?

Reply:
We need to differentiate `null` from lightweight binding from `null` from full binding, so that we can do the checks below and produce diagnostics based on the clauses from full binding only.
Moving this change down into `ResolveBounds` means that we would have multiple copies of that logic. Here seemed a more natural place (close to storage field).

Reply:
Many type parameters will have no constraints so it would be good to avoid calling `ResolveBounds()` above multiple times in those cases. Moreover, there are three implementations of `ResolveBounds(ConsList<TypeParameterSymbol> inProgress, bool canUseLightweightConstraintBinding, DiagnosticBag diagnostics)` that delegate to `ConstraintsHelper.ResolveBounds()` and one of them always sets `usedLightweightTypeConstraintBinding: false` so using `canUseLightWeightTypeConstraintBinding` here seems incorrect in that case at least.

Please consider moving the decision to return `NullFromLightweightBinding` to `ConstraintsHelper.ResolveBounds()`, or to the two implementations of `this.ResolveBounds()` here that require it.

---
In reply to: [499923263](https://github.com/dotnet/roslyn/pull/48270#discussion_r499923263) [](ancestors = 499923263)